DataFormats/simulation/include/SimulationDataFormat/MCCompLabel.h

```diff
@@ -48,53 +51,93 @@ class MCCompLabel
   static constexpr ULong64_t maskFull = (ul0x1 << (nbitsTrackID + nbitsEvID + nbitsSrcID)) - 1;
 
   MCCompLabel(int trackID, int evID = 0, int srcID = 0) { set(trackID, evID, srcID); }
-  MCCompLabel() = default;
+  MCCompLabel(unsigned int trackID, int evID, int srcID, bool fake) { set(trackID, evID, srcID, fake); }
+  MCCompLabel(bool invalid = false)
+  {
+    if (invalid) {
+      mLabel = Invalid;
+    }
+  }
   ~MCCompLabel() = default;
 
   // check if label was assigned
   bool isSet() const { return mLabel != NotSet; }
   // check if label was not assigned
   bool isEmpty() const { return mLabel == NotSet; }
-  // check if label was assigned with non-negaive trackID
-  bool isPosTrackID() const { return getTrackID() >= 0; }
-  // return 1 if the tracks are the same and have labels>=0,
-  // 0 if the tracks are the same but at least one of them has label<0
+  // check if label corresponds to real particle
+  bool isValid() const { return mLabel != Invalid; }
+  // check if label was assigned as for correctly identified particle
+  bool isSetAndValid() const { return isSet() && isValid(); }
+
+  // check if label was assigned as for incorrectly identified particle
+  bool isFake() const { return mLabel & Fake; }
+  // check if label was assigned as for correctly identified particle
+  bool isCorrect() const { return !isFake() && !(mLabel & Invalid); }
+
+  // return 1 if the tracks are the same and correctly identified
+  // 0 if the tracks are the same but at least one of them is fake
   // -1 otherwhise
   int compare(const MCCompLabel& other) const
   {
     if (getEventID() != other.getEventID() || getSourceID() != other.getSourceID()) {
       return -1;
     }
     int tr1 = getTrackID(), tr2 = other.getTrackID();
-    return (tr1 == tr2) ? 1 : ((tr1 == -tr2) ? 0 : -1);
+    return (tr1 == tr2) ? ((isCorrect() && other.isCorrect()) ? 1 : 0) : -1;
   }
 
   // conversion operator
   operator ULong64_t() const { return mLabel; }
   // allow to retrieve bare label
   ULong64_t getRawValue() const { return mLabel; }
 
-  // comparison operator, compares only label, not evential weight info
+  // comparison operator, compares only label, not evential weight or correctness info
   bool operator==(const MCCompLabel& other) const { return (mLabel & maskFull) == (other.mLabel & maskFull); }
   // invalidate
   void unset() { mLabel = NotSet; }
+  void invalidate() { mLabel = Invalid; }
+  void setFakeFlag(bool v = true)
+  {
+    if (v) {
+      mLabel |= Fake;
+    } else {
+      mLabel &= ~Fake;
+    }
+  }
+
   void set(int trackID, int evID, int srcID)
```

Comment:
I assume you kept this old interface for compatibility reason, but I would actually only provide the new setter with explicit fake flag. The automatic conversion of negative values to fakes is a bit misleading, now that we have an extra flag for it.

Reply:
ok, this required more changes...